### PR TITLE
Feature/56029 previous conversations view

### PR DESCRIPTION
**Pull request by fabclj**
This PR completes the Previous Conversations feature.
Related PR: #343 

It uses the last 5.0.0-beta.7 of socket-client, where it was added a new method `switchSession` we need for this feature.

It also introduces a concept for sync tabs, in case the user opens the widget on multiple tabs, on the same endpoint and same sessionId or a previuos conversation. We would need this because, so far, our chat updates the storage on every redux change, but never on the way back. The idea is to add the storage listener in order to update also redux store on inactive tabs. (details on footnote)
**EDIT**: this feature was removed on this PR. It can be included later by specific US, after agreement with PO.

Suggestions for testing:
- On this endpoint the bot answer with the user input text. It's useful to have different items in the list.
https://endpoint-dev.cognigy.ai/b18b807bba6602f42ba30feee3e76a587d69d034905059eb72ae15eaec57a3be

---
See also this page https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event
**Note**: The storage event is only triggered when a window **other than itself** makes the changes.

### CODE REVIEW

**Review by dshire — CHANGES_REQUESTED** [review]
So I read your code and tested it a bit, and while it works by itself, I don't know if we are breaking some obscure existing webchat feature here. We also reduce maintainability here with the state syncing by a lot. As this is an edge case scenario that was not part of the design documents, I am not super sure if we should add this logic now. Last week we already talked about this and we agreed that we shouldn't do it for now and should focus on the existing requirements, if I remember correctly?

We all agreed in the beginning to keep it simple to not risk breaking any existing logic. We need to stick to that paradigm to keep that risk as low as possible for now.

What do you guys think @fabclj @kwinto ?

**Comment by fabclj** [comment]
> So I read your code and tested it a bit, and while it works by itself, I don't know if we are breaking some obscure existing webchat feature here. We also reduce maintainability here with the state syncing by a lot. As this is an edge case scenario that was not part of the design documents, I am not super sure if we should add this logic now. Last week we already talked about this and we agreed that we shouldn't do it for now and should focus on the existing requirements, if I remember correctly?
> 
> We all agreed in the beginning to keep it simple to not risk breaking any existing logic. We need to stick to that paradigm to keep that risk as low as possible for now.
> 
> What do you guys think @fabclj @kwinto ?

It's fine to me to revert the changes about sync tabs logic. Question: should I comment the code with the idea to activate maybe later, or just I remove it at all?

**Comment by dshire** [comment]
> > So I read your code and tested it a bit, and while it works by itself, I don't know if we are breaking some obscure existing webchat feature here. We also reduce maintainability here with the state syncing by a lot. As this is an edge case scenario that was not part of the design documents, I am not super sure if we should add this logic now. Last week we already talked about this and we agreed that we shouldn't do it for now and should focus on the existing requirements, if I remember correctly?
> > We all agreed in the beginning to keep it simple to not risk breaking any existing logic. We need to stick to that paradigm to keep that risk as low as possible for now.
> > What do you guys think @fabclj @kwinto ?
> 
> It's fine to me to revert the changes about sync tabs logic. Question: should I comment the code with the idea to activate maybe later, or just I remove it at all?

If we revert it for now I would move it to a separate branch with very specific naming so that we can find it again easily and remove it from here. We could even create a new User Story and link the new branch there, then we can decide and track with the new User Story. Do you think that helps?

**Comment by fabclj** [comment]
> > > So I read your code and tested it a bit, and while it works by itself, I don't know if we are breaking some obscure existing webchat feature here. We also reduce maintainability here with the state syncing by a lot. As this is an edge case scenario that was not part of the design documents, I am not super sure if we should add this logic now. Last week we already talked about this and we agreed that we shouldn't do it for now and should focus on the existing requirements, if I remember correctly?
> > > We all agreed in the beginning to keep it simple to not risk breaking any existing logic. We need to stick to that paradigm to keep that risk as low as possible for now.
> > > What do you guys think @fabclj @kwinto ?
> > 
> > 
> > It's fine to me to revert the changes about sync tabs logic. Question: should I comment the code with the idea to activate maybe later, or just I remove it at all?
> 
> If we revert it for now I would move it to a separate branch with very specific naming so that we can find it again easily and remove it from here. We could even create a new User Story and link the new branch there, then we can decide and track with the new User Story. Do you think that helps?

I think we don't need a new branch. It could be enough to create a new US and link this PR.
It's not so much code, just the listener and a redux action. I would be able to reproduce it again with a few effort.

File: src/webchat-ui/components/presentational/previous-conversations/OverviewScreen.tsx
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat-ui/components/presentational/previous-conversations/OverviewScreen.tsx:1]: done

File: src/webchat-ui/components/WebchatUI.tsx [hunk 1]
```diff
@@ -383,6 +387,8 @@ export class WebchatUI extends React.PureComponent<React.HTMLProps<HTMLDivElemen
 
     renderInput = () => {
         const { inputPlugins } = this.state;
+        const { messages } = this.props;
+        if(isConversationEnded(messages)) return null;
```
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat-ui/components/WebchatUI.tsx]: fixed

File: src/webchat-ui/components/presentational/Chip.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	const { label, withWrapper, ...chipProps } = props;
+
+	if (!label) return null;
+
+	if (!withWrapper) return <ChipBase {...chipProps}>{label}</ChipBase>;
+
+	return (
+		<ChipWrapper>
+			<ChipBase {...chipProps}>{label}</ChipBase>
+		</ChipWrapper>
```
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat-ui/components/presentational/Chip.tsx:36]: added missing className

**Review by dshire — CHANGES_REQUESTED** [review]
Some more things I noticed:

1. We don't handle the case of "there are no previous conversations" right now. Maybe we should just disable the previous conversation button?
2. Back to overview functionality needs to be implemented with the new header (we should not forget it)
3. Do we already have a user story to prevent inputs from buttons in the chat history of a conversation that has ended? I think they probably need to be disabled

File: src/webchat/store/previous-conversations/previous-conversations-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+				const currentSession = store.getState().options.sessionId;
+				if (!currentSession) break;
+
+				const conversation = {
+					messages: store.getState().messages,
+					rating: store.getState().rating,
+				};
+				store.dispatch(upsertPrevConversation(currentSession, conversation));
+				break;
+			}
```
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat/store/previous-conversations/previous-conversations-middleware.ts:79]: yes, this one is not about sync tabs. We need to keep updated prevConversations state.

File: src/webchat-ui/components/presentational/Chip.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat-ui/components/presentational/Chip.tsx:36]: passed on [WebchatUI.tsx](https://github.com/Cognigy/WebchatWidget/pull/346/commits/771038559457f36bae4e9eb06547ea3cfd487d85#diff-3e22c3aa55508da4a4df11f22a9ad94f2d4721318ec54fdd8de8f66136858ad7)

We could reuse the Chip component for other cases, where a different className could be passed. What do you think?

File: src/webchat/store/previous-conversations/previous-conversations-middleware.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dshire) [reply src/webchat/store/previous-conversations/previous-conversations-middleware.ts:79]: let me rephrase it, why do we still need prevConversations state? can we not just load the data for the conversations list from localStorage when we need it?

**Comment by fabclj** [comment]
> Some more things I noticed:
> 
> 1. We don't handle the case of "there are no previous conversations" right now. Maybe we should just disable the previous conversation button?
> 2. Back to overview functionality needs to be implemented with the new header (we should not forget it)
> 3. Do we already have a user story to prevent inputs from buttons in the chat history of a conversation that has ended? I think they probably need to be disabled

I answer by points:

1. In US story was mentioned to keep the list empty in this case.
2. This was already discussed in chat with @kwinto , we should have a simple 'navigation history' implemented in redux UI store.
3. This is a good point, afaik this case was not mention anywhere.

File: src/webchat/store/previous-conversations/previous-conversations-middleware.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat/store/previous-conversations/previous-conversations-middleware.ts:79]: Actually it could be possible to work directly with localStorage, but I personally prefer to handle the data we are consuming in the components using state instead. It looks more consistent and flexible to me.

**Review by kwinto — APPROVED** [review]
lgtm

When we will be adding expiry based on endpoint config, we should also look into
blocking sending of the messages to the expired sessions with `sendMessage` Webchat API.

File: src/webchat/components/Webchat.tsx [hunk 4]
```diff
@@ -56,12 +57,28 @@ export class Webchat extends React.PureComponent<WebchatProps> {
...
+        }
     }
 
     componentWillUnmount() {
         this.client.disconnect();
+        window.removeEventListener('storage', this.syncStorage);
+    }
+
+    syncStorage = (e: StorageEvent) => {
+        if (e?.key && e?.newValue) {
```
[the remark this answers is not in the document]
Comment (fabclj) [reply src/webchat/components/Webchat.tsx]: > The Webchat will be run alongside other applications, which may also use localStorage. We should add some check here to dispatch only on the related events.

In this case the Storage event is only one. The only we can do is to filter it in order to be sure that the modified key/value pair is a record we need to handle. This kind of check was already added on the previos implementation. We will discuss eventually how to improve it, if needed, when we will be back on it.

**Comment by fabclj** [comment]
> lgtm
> 
> When we will be adding expiry based on endpoint config, we should also look into blocking sending of the messages to the expired sessions with `sendMessage` Webchat API.

I agree, this should be done. But I also think we should disable all the buttons on the chat history.
